Add Gmail Tool

1239uth opened the pull request:
# Summary
This tool adds support for Google's Gmail API so that Agents can interact with a user's Gmail account to 
- Create drafts
- Get messages
- Get threads
- Search for messages/threads
- Send messages

There are equivalent features already implemented in the Python version ([link to python tools/gmail](https://github.com/langchain-ai/langchain/tree/f5326cfb4e10b3b0dd6cac1f51c56616ba931c19/libs/langchain/langchain/tools/gmail))

This follows closely with the implementation of the Google Calendar tool ([link to tools/google_calendar](https://github.com/langchain-ai/langchainjs/tree/f3c2f169c2d2d2e7565ea37496260fd6206f9f4f/langchain/src/tools/google_calendar))

# Code
- Added class `GmailBaseTool extends StructuredTool` that initializes the gmail object with the env variables for google auth
- Added classes `GmailCreateDraft`, `GmailGetMessage`, `GmailGetThread`, `GmailSearch`, `GmailSendMessage` which extend `GmailBaseTool`
- Added docs with example implementation (similar to google calendar)

# Testing

- Added unit tests for input handling logic for various inputs in the constructor for GmailBaseTool
- Was difficult to test google JWT auth since we needed a google workspace platform account, but we know from running the example that the tools are being called correctly and an API request to google is going out.

# TODO
- We didn't get time to implement the GmailToolkit

Looking forward to receiving feedback on our implementation!

We'd like to give credit to the University of Toronto for making us do our first open source contribution!

Fixes #2946

1239uth commented:
We are unable to see the build failure logs for the Vercel docs deployment. Would appreciate help on how to fix that

1239uth commented on langchain/src/tools/gmail/send_message.ts:
```diff
@@ -0,0 +1,74 @@
+    return Buffer.from(email).toString("base64url");
+  }
+
+  async _call({
```
It doesn't seem like `_call` it should be overriding anything but when we run the example file it seems to correctly call the method and make the API request to gmail.
Also we used `Tool` because it extends `StructuredTool` already, and google calendar was also doing the same thing. But on second thought, we aren't overriding the `call` method defined in `Tool` anyways, so maybe that's pointless. We can make the change to make `GmailBaseTool extend StructuredTool` instead.

1239uth replied:
On second thought, it seems like a `Toolkit` requires a list of `Tool` only, not `StructuredTool`. https://github.com/langchain-ai/langchainjs/blob/d82cfb84ac60367595d19e77175c7f1c596c247c/langchain/src/agents/toolkits/base.ts#L8-L10

So making these of type `StructuredTool` would mean we cannot make a Toolkit with the Gmail tools right? In that case is it better to keep these as a `Tool`?

jacoblee93 replied:
Yeah, we can modify Toolkit or I suppose make a `StructuredToolkit` in the future but these should definitely extend `StructureTool`

1239uth replied:
Alright that makes sense. We've made the change to convert `GmailBaseTool` into a `StructuredTool` already

1239uth commented on langchain/src/tools/gmail/create_draft.ts:
```diff
@@ -0,0 +1,65 @@
+import { GmailBaseTool, GmailBaseToolParams } from "./base.js";
```
The python version had create_draft so we wanted to allow the same tools available there. I'm guessing creating a draft might be useful because you may not fully trust the LLM to write AND send an email for you so you just want it to create a draft for you that you can potentially edit later?

Up to you though, if you don't feel like this is necessary we can delete it.

1239uth commented on docs/core_docs/docs/integrations/tools/gmail.mdx:
```diff
@@ -0,0 +1,23 @@
+
+## Setup
+
+To use the Gmail Tool you need to install the following official peer dependency:
```
Sure, adding that.

1239uth commented:
We've addressed the comments and fixed the docs build issue. But for some reason when I try to resolve the merge conflicts by pulling main into this branch it causes a build failure that I wasn't able to fix:
```
src/chat_models/minimax.ts:1:35 - error TS2307: Cannot find module '@langchain/openai' or its corresponding type declarations.

1 import type { OpenAIClient } from "@langchain/openai";
                                    ~~~~~~~~~~~~~~~~~~~

Found 1 error in src/chat_models/minimax.ts:1
```
Would appreciate some help on fixing that.

jacoblee93 commented:
Hey @1239uth, seems like your branch is protected and I can't push to it :(

Mind either changing that or merging `main` yourself?

1239uth commented:
> Hey @1239uth, seems like your branch is protected and I can't push to it :(
> 
> Mind either changing that or merging `main` yourself?

Sorry @jacoblee93, removed the branch protection rule

1239uth commented on examples/src/tools/gmail.ts:
```diff
@@ -0,0 +1,56 @@
+
+  const createInput = `Create a gmail draft for me to edit of a letter from the perspective of a sentient parrot who is looking to collaborate on some research with her estranged friend, a cat. Under no circumstances may you send the message, however.`;
+
+  const createResult = await gmailAgent.call({ input: createInput });
```
Sounds good, made that change

1239uth commented:
> Hey @1239uth, seems like your branch is protected and I can't push to it :(
> 
> Mind either changing that or merging `main` yourself?

Merged main, should be good now

jacoblee93 commented:
Thank you!

alexleventer commented:
I haven't been able to get auth to work
